05-tasks-and-events.md

```diff
@@ -34,85 +41,134 @@ Tasks MUST follow those rules:
+    - For transaction mined, emit `(tx_block, Some(confirmations))` where `tx_block` is the block that the transaction got mined, and `confirmations` is the number of blocks extending `tx_block`
+    - When `confirmations` >= `confirmation_bound`, emit `(tx_block, Some(confirmation_bound))`. Here `Task` is considered successfully accomplished and terminates. At that time, transaction is considered final (=irreversible). 
+
+Thus:
+```
+for each transaction being watched by Syncer:
+  for every new block arrival extending the chain that the transaction was mined: 
+    if confirmations <= confirmation_bound:
+      Syncer emits: `(tx_block, Some(confirmations))`
+```
+
+Upon block reorg reverting transaction out of blockchain, emit `(0x0, Some(0))` if transaction shows back in mempool or `(0x0, None)` if it's not in the mempool.
+
+### The `transaction_broadcasted` Event
+The `broadcast_transaction` task defined above produces upon successful transaction broadcast a success output message: `transaction_broadcasted` blockchain event.
+
+This message contains:
+* `tx`: The raw transaction in its serialized format.
+* `success_broadcast`: bool
+
 
-Blockchain Events are produced by syncers in response to certain type of `tasks`. A task MAY produce multiple `blockchain events`. When a task produces two different set of blockchain events depending on when the task is handled by the syncer those two sets MUST have an equivalent impact on the state at any point in time.
+FIXME: move to state RFC
+### Equivalence of Event Sets 
 
-### Equivalence of Event Sets
-Let's define a `new height` task, this task produces `height changed` blockchain events upon new block and reorgs. Let's define `X` as the current block height. The task is sent to the syncer, initial plus two events are received, for `X`, `X+1`, and `X+2` new heights. At time `t` the latest state is for `X+2`. If the daemon crashes at `X+1` and restarts, at time `t` it MUST have received `X+1` as initial event and `X+2`, the latest state is the same. And finally if the daemon crashes at time `t` and restarts, the initial blockchain event MUST contains `X+2`. Sets of events are different but equivalent for the daemon state.
+When a task produces two different sets of blockchain events depending on when the task is handled by the syncer those two sets MUST have an equivalent impact on the state of the daemon at any point in time.
 
-Let's define a `broadcast transaction` task for tasks that have side effects, this job produces as a success output a `transaction broadcasted` blockchain event. The daemon sends the task at time `t` and recieves the successful `transaction broadcasted` event at time `t'`, if the daemon crashes between `t` and `t'`, rebroadcasting the task MUST result in the same successful event, despite the fact that the syncer may not broadcast the transaction to the full-node a second time.
+Examples follows.
 
-### The `tada` Event
+#### `block_height` event
```

Comment:
This is a really complex explanation that is difficult to ack.

Let `X` be the current blockchain height and `t1` be the time at the start. Two new blocks appear, yielding height events`X+1` and `X+2`. `t2` is the current time.

I planned to continue with a suggestion, yet this is where I got lost. This says the daemon received X+1 (when it crashed) and then X+2, yet it crashed before X+2. Are we assuming it can hook into existing communication channels and catch up? Or are we trying to say it'll need to get the X+2 event? Or that the X+2 event doesn't matter?

Reply:
I think this is @h4sh3d work, i propose we later move it to the state rfc, because this problem applies to other components (syncer and client)